cmd/nvidia-device-plugin/main.go

```diff
@@ -225,6 +226,31 @@ restart:
 				klog.Infof("inotify: %s created, restarting.", pluginapi.KubeletSocket)
 				goto restart
 			}
+			if event.Name == mps.ReadyFilePath {
+				if config == nil || config.Sharing.SharingStrategy() != spec.SharingStrategyMPS {
+					klog.InfoS("Ignoring event", "event", event)
+					continue
+				}
+				switch {
+				case event.Op&fsnotify.Remove == fsnotify.Remove:
+					klog.Infof("%s removed; restarting", mps.ReadyFilePath)
+					goto restart
+				case event.Op&(fsnotify.Create|fsnotify.Write) != 0:
+					klog.Infof("%s created or modified; checking config", mps.ReadyFilePath)
+					readyFile := mps.ReadyFile{}
+
+					matches, err := readyFile.Matches(config)
+					if err != nil {
+						klog.ErrorS(err, "failed to check .ready file")
+						goto restart
+					}
+					if !matches {
+						klog.Info("mismatched MPS sharing config; restarting.")
+						goto restart
+					}
+					continue
+				}
+			}
```

Comment:
I haven't read through this in detail, but at first glance it seems a bit convoluted. Can't we rely on the logic already in-place that handles config updates and hook into that somehow?

Reply:
It may be possible to just rely on the config change restart logic. Let's make the following assumptions:

1. The MPS daemonset owns the lifecycle of the `.ready` file and this file contains the current config used to start the MPS control daemon.
2. If the config changes and both the MPS daemon and the device plugin are restared by their respective config managers, we should be able to fail plugin startup until: 1. A .ready file exists with the expected config, 2. The MPS pipe is ready from the perspective of the device plugin.

This should not require the additional error / watcher events here.

Note that since the `config-manager` sidecars are not always deployed, this seems to imply that there are configuration changes that we may miss in this case. For example, how do we capture the restart of the MPS daemon or some other configuration update that doesn't use the config manager? Do we expect the user to manually restart both pods in this case to ensure that we're in sync?